LRA extension blog post

mmusgrov opened the pull request:
This blog should coincide with the release of the Narayana LRA extension (https://github.com/quarkusio/quarkus/pull/17903) which will be in the 2.2 release.

ochaloup commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+
+Now update the generated application so that the hello method will execute in the context of a Long Running Action:
+
+Open the file `src/main/java/org/acme/GreetingResource.java` in an editor and annotate the hello method with an @LRA annotation (also inject the LRA context into the method using the JAX-RS `HeaderParam` annotation). In addition, add two callback methods which will be called when the LRA is closed or cancelled.
```
```suggestion
Open the file `src/main/java/org/acme/GreetingResource.java` in an editor and annotate the hello method with an `@LRA` annotation (also inject the LRA context into the method using the `@javax.ws.rs.HeaderParam` annotation). In addition, add two callback methods which will be called when the LRA is closed or cancelled.
```

mmusgrov replied:
ok

ochaloup commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+[1] 2495275
+----
+
+Start an LRA using curl to send a request the new method we have just added:
```
```suggestion
Start an LRA using `curl` to send a request the new method we have just added:
```

mmusgrov replied:
ok

ochaloup commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+
+== Native mode
+
+NOTE: in native mode only external coordinators (i.e. not embedded with the application) are supported (we will provide a coordinator extension in a later release to address this deficiency).
```
The text above contains similar remark
(`(NOTE: you cannot use this approach to run coordinators in native mode, a future extension will be provided to support this requirement).`)
isn't this one an duplication?

mmusgrov replied:
The note concerns running coordinators in native mode and it is important to include it in the native mode section and to also  prefix it with an asciidoc `NOTE:`. Therefore if I remove the initial reference, in the section on building coordinators, then the user may assume that he can produce native coordinators, and only if he reads the later section will he realise that his initial assumption was incorrect. Furthermore, if the user came to the section via a google search then he would potentially miss it altogether).

Secondly, the second note is referring to embedded coordinators whereas the first note is referring to building standalone native coordinators and therefore it is subtly different from the second reference.

ochaloup replied:
Ok, I see.

ochaloup commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+The same caveats as described in the <<_lra-coordinators,the section on coordinators>> still apply:
+
+- no support for native executables.
+- each instance requires dedicated storage for its transaction logs (since sharing transaction stores is not supported).
```
Do we consider starting two coordinators in the same app or a participant and coordinator.
If the later then the participant does not require to work with the transaction log storage.  And maybe the second caveat is not needed to be mentioned?

mmusgrov replied:
Two coordinators in one app does not make much sense (and I cannot conceive of a use case).

But a coordinator embedded with a participant does make sense and the associated storage does need to be managed: the fact that a participant is running in the same JVM as the coordinator makes no difference to the fact that the storage needs to be dedicated for use by only that coordinator.

ochaloup replied:
I see, my point was that the participant does not need any storage that would be related to Narayana. The object store is used only by coordinator. The participant does not touch it. Participant will usually have own "unrelated" storage.

The note is not problematic just it seemed to me to hint that the participant could be somehow related to object store which should not be so.

mmusgrov replied:
Your point is noted but embedding a coordinator with the participant implies that the associated storage for the transaction logs cannot be shared. I reworded the sentence to make it clear:

`dedicated storage for its transaction logs` -> `dedicated storage for the coordinators' transaction logs`

ochaloup reviewed (commented):
The text is nice. I have got just few notes.

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+---
+layout: post
+title: An Extension for Long Running Activities
```
Is activity not already a well understood term? Activity and Action are often used interchangeably in computing and if using them that way does not hinder the users understanding of LRA then I do not see any downside to using the words interchangeably  (in fact it encourages a more expansive view of what the specification is providing to business application developers since Activity probably encompasses the definition of Action).

ochaloup replied:
Within the LRA we use the most the term `Action`. I understand it can be used interchangeably. My point was that would be good to be consistent to naming used in LRA spec.
From what you say I understand you prefer the term `Activity` here. That's fine with me.

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+
+We call the interaction an LRA, short for Long Running Action. An LRA has specific properties and guarantees that aid the construction of reliable service interactions. Each action is assigned a unique identifier (referred to as the LRA context) so that it can be distinguished from other LRAs.
+
+Services start an LRA by marking a JAX-RS method with an https://download.eclipse.org/microprofile/microprofile-lra-1.0/apidocs/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.html[@LRA annotation]. When such a method is invoked the system will start the action and make its identifier available as a JAX-RS header called "Long Running Action". If the body of the method performs any JAX-RS invoca ...
```
It would indeed, agreed.

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+[source,bash]
+----
+$ mvn io.quarkus:quarkus-maven-plugin:2.2.0.Final:create \
+      -DprojectGroupId=org.acme \
```
Most quickstarts use `org.acme` and most user applications will want to use their own group ids.

ochaloup replied:
+1

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+      -DprojectArtifactId=narayana-lra-coordinator \
+      -Dextensions="resteasy-jackson,rest-client"
+$ cd narayana-lra-coordinator/
+$ rm -rf src
```
Will do.

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+Towards the end of article we will show how to embed coordinators with services
+(NOTE: you cannot use this approach to run coordinators in native mode, a future extension will be provided to support this requirement).
+
+== Non native mode
```
I was wanting to draw out the distinction (between native and JVM) but I guess I can reword it if you think it's confusing.

ochaloup replied:
I was just thinking that the Quarkus usually use the term "JVM mode". My point was to be consistent in the usual naming around Quarkus. I do understand both. If the current term shows better the point it can be used as it is.

mmusgrov replied:
I have gone with your suggestion.

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+
+To build a JAX-RS application with LRA support add the dependency `io.quarkus:quarkus-narayana-lra` to your application's pom. This will add JAX-RS support and make the LRA annotations available when developing your services, it also register a JAX-RS filter that automatically manages the participation of your services in LRAs.
+
+The guarantees (of eventual consistency) required by the https://download.eclipse.org/microprofile/microprofile-lra-1.0-M1/microprofile-lra-spec.html[LRA specification] rely on the presence of a JAX-RS application that coordinates the services participating in the LRA. This component must be present when starting the interaction, when joining the interaction and when ending the interaction. If th ...
```
The system is the infrastructure and set up that implements the LRA specification so I'm happy with the term system.

ochaloup replied:
Ok.

mmusgrov commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+
+[source,bash]
+----
+$ quarkus.lra.coordinator-url=http://localhost:<port>/lra-coordinator
```
The next lines already indicate that the reader should rebuild the app, quote:

```
Verify that the generated application still works after these changes:

[source,bash]
----
$ ./mvnw clean package
----
```

and the reference to updating the config is just a *note* to say that if your coordinator is not on the default port then the developer should specify where he is running coordinators; Quarkus config is preferable to java system properties when the developer knows up front what values to use.

ochaloup replied:
Ok.

gsmet commented:
@mmusgrov please ping me on Zulip when you want this merged. Also, be careful, update the two dates: the one in the file name and the one in the adoc. Thanks.

gsmet commented on _posts/2021-08-23-using-lra.adoc:
```diff
@@ -0,0 +1,490 @@
+
+[source,bash]
+----
+$ mvn io.quarkus:quarkus-maven-plugin:2.2.0.Final:create \
```
Please make it 2.2.1.Final, we won't officially announce 2.2.0.Final due to a bug in dev mode on Windows.

mmusgrov commented:
@ochaloup I pushed an update which addresses your comments (and I made a few minor tweaks to the grammar). Let me know if it's satisfactory so that I can ping Guillaume?

ochaloup reviewed (approved):
Thanks,  it's fine.

gsmet commented:
I adjusted the date, added a mention in the announcement and added you to the authors.yaml file (if you want to picture, you need to add the gravatar hash there).

mmusgrov commented:
> I adjusted the date, added a mention in the announcement and added you to the authors.yaml file (if you want to picture, you need to add the gravatar hash there).

Thanks for the modifications @gsmet 

I already synced the date at the top of the blog file and the name of the blog file (they're both 2021-08-23). Where did you adjust the date?

Also my original PR included  a change to `authors.yaml` including bio, job title etc. What you've put is good but I am puzzled about what happened to my original update to this file.

gsmet commented:
I don’t know but it wasn’t in your final commit. I had nothing after By in your blog post thus why I quickly added some basic info. 

I adjusted the date because, well, we are September 1st now and 2.2 was announced yesterday.
